### ISSUE TITLE
MacPass permission to record screen not working with Pulse Secure

### ISSUE THREAD

**Issue by chipmunk404** [MacPass/MacPass#992]
I just installed the newest iteration of this wonderful piece of software, hoping that Autotype would work again on Catalina. For the most part, it does work, however when I want to Perform Autotype in Pulse Secure (a VPN software), MacPass says that the right permissions are not set.
Is this expected behaviour? On websites the autotype does work.

![image](https://user-images.githubusercontent.com/57216165/67936485-2722c100-fbcc-11e9-8925-0ce111487df8.png)

 - macOS: 10.15.0
 - Version 0.7.10

**Comment by georgesnow** [comment]
I have had issues with pulse myself and auto type, but in my testing it seems related to the fact the window title And window becoming active. Curious are using global auto type with the hot key or triggering auto type from within macpass?

**Comment by mstarke** [comment]
@chipmunk404 do you get the permission message only, when you use Global Autotype for Pulse Secure? Or do you get it on every invocation of Global Autotype?

**Comment by chipmunk404** [comment]
@georgesnow @mstarke The same applies to both ways of doing autotype, both by using the Global Autotype shortcut and by clicking Autotype from within MacPass.
So far, this only applies to Pulse Secure, websites and so far other apps seem to work just fine.

**Comment by georgesnow** [comment]
So with pulse secure that slide out window causes the window name to be obscured in my testing. And only the active application name is available which also causes an issue to make the pulse window active

**Comment by georgesnow** [comment]
Window becomes active but the cursor won’t land back in the username box even if it was there when triggered

**Comment by mstarke** [comment]
MacPass performs Autotype by switching to the last active application, so if the input does not gain focus again on re-activation that's the cause for Autotype not working. Global Autotype however does not de- or re-activate an app but simply pushes keyboard events. Pulse should not loose focus and the input field should receive the events. It might be possible that the input fields do not play well with the way MacPass sends the key strokes.

**Comment by mstarke** [comment]
I've just discovered a flaw in the Autotype doctor that determines if screen recording permissions are granted. Each time global autotype is invoked, the Autotype doctor determines the number of window names it can read as this seemed to be a meaningfull way of checking for screen recording permissions without having to actually capture any pixel on the screen. This is done every time so there may be edge cases (like the Pure client) where the way to determine if screen recording are set is a false negative. I have to dig into this some more.

**Comment by georgesnow** [comment]
@mstarke correct, sorry **_global_** auto type works for Pulse Secure, but "Auto type" that you trigger from MacPass Pulse doesn't due to the focus issue. Side note in 10.15.1 Accessibility list won't let you uncheck MacPass. 
![output](https://user-images.githubusercontent.com/12751441/67954377-3d7a4e00-fbc7-11e9-91a4-da7b0e03b4eb.gif)

**Comment by mstarke** [comment]
@georgesnow Thank you for the insights. Could you please open a new issue regarding the accessibilty problems? I do want to prevents issue being derailed into multi-issue discussion.

**Comment by mstarke** [comment]
I've changed the behaviour on how the screen recording permissions are determined which should fix the false alert. This fix will be included in 0.7.11. However this seems not to be the root problem, as the client does not gain focus on the input field upon re-activation. This is rather unusual and seem impossible to be fixed since this would require app specific code to re-focus input elements which is quite the opposite of what Global Autotype tries to achieve - a simple universal way.

**Comment by georgesnow** [comment]
well this stupid saga continues. apparently pulse secure is able to block auto type. I can get auto type to work everywhere else, but as soon as I try in the pulse secure app it causes MacPass to say it doesn't have the permissions anymore. It seems I can get it to break it every time....errr. it seems like an app can have sort of flag maybe?

**Comment by georgesnow** [comment]
and as long as one of the windows is up in the workspace it blocks auto type. however, once you close all its windows. auto type immediately starts working again. Just as reference here Pulse Secure is a menu bar app (ie ..NSStatusItem) and runs without a dock icon (ie LSUIElement)

**Comment by georgesnow** [comment]
one other tidbit this started happening when I upgraded macOS 10.15.1 (I haven't changed anything else). 10.15.0 worked the way I described above. So it seems like Pulse Secure somehow blocks apps that has screen recording permissions and the dot update it is honoring that maybe??!
The thing is I don't think I have any other apps that use screen recording to test against.

**Comment by georgesnow** [comment]
yeah, cause this bothers regularly enough. I dug around some more. pulse must be monitoring (looking for posts to the notification center) for NSNotifications (I believe)for screen recording. I tried to record with QuickTime the window and took screenshot and killed the screen video recording reproducible  every time . Cause Autotype (not global) works. I couldn't figure out hot (or find documentation on how to monitor for screen recording). but it is possible:

https://objective-see.com/blog/blog_0x2E.html

I tested and this works for all sorts of things like the lock screen, which is not posted the general UI. However, it is a posted notification. anyway, this is all to say still no workaround for Global Autotype within the main app... 

I have just been using my development plugin, which matches on bundle id and I trigger Autotype that way.
